File: tests/fixtures/selenium-chaos-suite/src/pages/ProductsPage.ts
import { By } from "selenium-webdriver";
import { BasePage } from "./BasePage";
import { AssertionHelper } from "../core/AssertionHelper";
import { WaitUtils } from "../core/WaitUtils";

export class ProductsPage extends BasePage {
  async assertInventoryPageLoaded() {
    const title = await this.getTextByCss(".title");
    AssertionHelper.shouldEqual(title, "Products", "Inventory title mismatch");
  }

  async sortLowToHigh() {
    await this.clickByCss(".product_sort_container");
    await this.driver.findElement(By.css(".product_sort_container")).sendKeys("lohi");
    await WaitUtils.sleep(500);
  }

  async sortHighToLow() {
    await this.clickByCss(".product_sort_container");
    await this.driver.findElement(By.css(".product_sort_container")).sendKeys("hilo");
    await WaitUtils.sleep(500);
  }

  async addBackpackToCart() {
    await this.clickByCss("#add-to-cart-sauce-labs-backpack");
    const badge = await this.getTextByCss(".shopping_cart_badge");
    AssertionHelper.shouldEqual(badge, "1", "Cart badge should be 1");
  }

  async addBikeLightToCart() {
    await this.clickByCss("#add-to-cart-sauce-labs-bike-light");
    const badge = await this.getTextByCss(".shopping_cart_badge");
    AssertionHelper.shouldContain(badge, "2", "Cart badge should contain 2");
  }

  async addAllItemsOneByOne() {
    const addButtons = await this.driver.findElements(By.css("button.btn_inventory"));
    for (const btn of addButtons) {
      await btn.click();
      await WaitUtils.sleep(100);
    }
    const badge = await this.getTextByCss(".shopping_cart_badge");
    AssertionHelper.shouldEqual(badge, String(addButtons.length), "Badge should equal added items");
  }

  async openFirstItemAndAssertDetails() {
    await this.clickByCss(".inventory_item_name");
    const name = await this.getTextByCss(".inventory_details_name");
    const desc = await this.getTextByCss(".inventory_details_desc");
    const price = await this.getTextByCss(".inventory_details_price");
    AssertionHelper.shouldContain(name.toLowerCase(), "sauce", "Unexpected item name");
    AssertionHelper.shouldBeTruthy(desc.length > 10, "Description too short");
    AssertionHelper.shouldContain(price, "$", "Price should include currency");
  }

  async goToCartFromHeader() {
    await this.clickByCss(".shopping_cart_link");
    await this.assertUrlContains("cart");
  }

  async stressOpenAndBack(rounds = 6) {
    for (let i = 0; i < rounds; i++) {
      const items = await this.driver.findElements(By.css(".inventory_item_name"));
      if (!items.length) {
        break;
      }
      await items[0].click();
      await WaitUtils.sleep(200);
      await this.driver.navigate().back();
      await WaitUtils.sleep(200);
    }
    const body = await this.driver.findElement(By.css("body")).getText();
    AssertionHelper.shouldContain(body, "Products", "Expected to return to Products page");
  }
}

File: tests/fixtures/selenium-chaos-suite-csharp/src/Pages/ProductsPage.cs
using OpenQA.Selenium;

namespace SeleniumChaosSuite.Pages;

public class ProductsPage : BasePage
{
    public ProductsPage(IWebDriver driver) : base(driver) { }

    public Task<string> GetPageTitle()
    {
        return GetTextByCss(".title");
    }

    public async Task SortLowToHigh()
    {
        await ClickByCss(".product_sort_container");
        var sort = Wait.Until(d => d.FindElement(By.CssSelector(".product_sort_container")));
        sort.SendKeys("lohi");
        Wait.Until(d => d.FindElements(By.CssSelector(".inventory_item")).Count > 0);
    }

    public async Task SortHighToLow()
    {
        await ClickByCss(".product_sort_container");
        var sort = Wait.Until(d => d.FindElement(By.CssSelector(".product_sort_container")));
        sort.SendKeys("hilo");
        Wait.Until(d => d.FindElements(By.CssSelector(".inventory_item")).Count > 0);
    }

    public async Task<string> AddBackpackToCart()
    {
        await ClickByCss("#add-to-cart-sauce-labs-backpack");
        return await GetTextByCss(".shopping_cart_badge");
    }

    public async Task<string> AddBikeLightToCart()
    {
        await ClickByCss("#add-to-cart-sauce-labs-bike-light");
        return await GetTextByCss(".shopping_cart_badge");
    }

    public async Task<(string badge, int count)> AddAllItemsOneByOne()
    {
        var addButtons = Driver.FindElements(By.CssSelector("button.btn_inventory"));
        foreach (var btn in addButtons)
        {
            btn.Click();
        }

        Wait.Until(d => d.FindElement(By.CssSelector(".shopping_cart_badge")).Displayed);
        var badge = await GetTextByCss(".shopping_cart_badge");
        return (badge, addButtons.Count);
    }

    public async Task OpenFirstItem()
    {
        await ClickByCss(".inventory_item_name");
    }

    public async Task<(string name, string description, string price)> GetItemDetails()
    {
        var name = await GetTextByCss(".inventory_details_name");
        var desc = await GetTextByCss(".inventory_details_desc");
        var price = await GetTextByCss(".inventory_details_price");
        return (name, desc, price);
    }

    public async Task AddBoltShirtToCart()
    {
        await ClickByCss("#add-to-cart-sauce-labs-bolt-t-shirt");
    }

    public async Task AddJacketToCart()
    {
        await ClickByCss("#add-to-cart-sauce-labs-fleece-jacket");
    }

    public async Task AddOnesieToCart()
    {
        await ClickByCss("#add-to-cart-sauce-labs-onesie");
    }

    public async Task AddRedShirtToCart()
    {
        await ClickByCss("#add-to-cart-test.allthethings()-t-shirt-(red)");
    }

    public async Task RemoveBackpackFromCart()
    {
        await ClickByCss("#remove-sauce-labs-backpack");
    }

    public async Task RemoveBikeLightFromCart()
    {
        await ClickByCss("#remove-sauce-labs-bike-light");
    }

    public Task SortBy(string value)
    {
        var sort = Wait.Until(d => d.FindElement(By.CssSelector(".product_sort_container")));
        sort.SendKeys(value);
        Wait.Until(d => d.FindElements(By.CssSelector(".inventory_item")).Count > 0);
        return Task.CompletedTask;
    }

    public Task OpenItemByIndex(int n)
    {
        var items = Driver.FindElements(By.CssSelector(".inventory_item_name"));
        if (n >= 0 && n < items.Count)
        {
            items[n].Click();
            Wait.Until(d => d.FindElement(By.CssSelector("body")).Displayed);
        }

        return Task.CompletedTask;
    }

    public async Task GoToCartFromHeader()
    {
        await ClickByCss(".shopping_cart_link");
        await AssertUrlContains("cart");
    }

    public Task<string> StressOpenAndBack(int rounds = 6)
    {
        for (var i = 0; i < rounds; i++)
        {
            var items = Driver.FindElements(By.CssSelector(".inventory_item_name"));
            if (items.Count == 0)
            {
                break;
            }

            items[0].Click();
            Wait.Until(d => d.FindElement(By.CssSelector("body")).Displayed);
            Driver.Navigate().Back();
            Wait.Until(d => d.FindElements(By.CssSelector(".inventory_item_name")).Count > 0);
        }

        return Task.FromResult(Driver.FindElement(By.CssSelector("body")).Text);
    }
}
